### PR TITLE
MNG-5889 - adding logic that looks for the file argument and starts t…

### PR DESCRIPTION
**Pull request by rpatrick00**
…he search for the .mvn directory at the location of the specified POM when present

### CODE REVIEW

**Comment by Tibor17** [comment]
@rpatrick00 
It looks like "--file" is not recognize but "-file" is recognized.
The second issue where I have `/a/b` folders and `b` contains single `pom.xml` and `.mvn` - no multimodule.
`cd /a` which does not have `.mvn` and start `mvn -file b/pom.xml`. There my `vm.config` is not applied, why? Tested on Win. I will try in Fedora as well.

**Comment by rpatrick00** [comment]
OK, I found and resolved the problem with the --file arg but am not seeing the other problem you describe (see output below).  Maybe my fix for the --file arg resolved that one too?  

Here is a stupid little test project that I am using:
[maven-test.zip](https://github.com/apache/maven/files/518411/maven-test.zip)

In the root directory of the project, there is a git.diff file that shows the changes I made to the pull request code.  I am happy to resubmit the pull request with this change but would prefer to verify that it fixes the issues in your Windows environment first.

If you run with JDK 8, you will see the warning about ignoring MaxPermSize=128m, which is set in jvm.config:

D:\temp>mvn --file maven-test\pom.xml verify
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=128m; support was removed in 8.0
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building maven-config-test 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- echo-maven-plugin:0.3.0:echo (default) @ maven-config-test ---
[INFO]
[INFO]
[INFO]
[INFO]
[INFO] myproperty = from.maven.config
[INFO]
[INFO]
[INFO]
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 0.587 s
[INFO] Finished at: 2016-10-09T11:20:56-05:00
[INFO] Final Memory: 17M/491M
[INFO] ------------------------------------------------------------------------

**Comment by Tibor17** [comment]
Don't create a new PR. Just amend the last commit `git commit --amend` and push it again.

**Comment by rpatrick00** [comment]
Pushed

**Comment by Tibor17** [comment]
@rpatrick00 
LGTM
I have tested the change on Win7 and Fedora23.
I will merge your PR with master.

File: apache-maven/src/bin/mvn.cmd [hunk 1]
```diff
@@ -86,19 +86,65 @@ set MAVEN_CMD_LINE_ARGS=%*
...
+) else (
+  goto findBaseDir
+)
+
+:process_file_arg
+if "%FILE_ARG%" == "" (
+  goto findBaseDir
+)
+call :get_directory_from_file %FILE_ARG%
+if not exist "%POM_DIR%" (
```
Comment (michael-o) [inline apache-maven/src/bin/mvn.cmd]: Why is this check not present in the shell script?

File: apache-maven/src/bin/mvn.cmd [hunk 2]
```diff
@@ -86,19 +86,65 @@ set MAVEN_CMD_LINE_ARGS=%*
...
+  echo Directory %POM_DIR% extracted from the -f/--file command-line argument does not exist
```
Comment (michael-o) [inline apache-maven/src/bin/mvn.cmd]: This should go to `stderr`.

Comment (rpatrick00) [inline apache-maven/src/bin/mvn.cmd]: OK

**Comment by rpatrick00** [comment]
OK, addressed both comments from michael-o

**Comment by Tibor17** [comment]
@michael-o 
Any objections to merget this or do we need to test it with specific use case?

**Comment by michael-o** [comment]
I think this really deserves an IT too. The shell script is fine. Can't just on the command script, I am virtually illiterate here. I trust @rpatrick00  here. Willing to blindly to merge when the IT is ready.

**Comment by Tibor17** [comment]
@michael-o 
I will merge it no problem. The only problem on IT is my spare time.

**Comment by michael-o** [comment]
I would rather see @rpatrick00 providing the IT.

**Comment by rpatrick00** [comment]
@michael-o, while I have no problem contributing, I have no familiarity with the maven-integration-test project so it will likely take me some time to reverse engineer it enough to figure out how to create an integration test for the shell scripts and the various use cases that should be tested.  I would hate to see this fix gated by the requirement for me to accomplish this given the demands of my day job...

**Comment by hboutemy** [comment]
let's help: I dug into core-its, to find what existing IT to extend, since .mvn feature should have an IT already
=> found mng-5771-core-extensions IT, which already tests 3 conditions: I think it's the right place to add one more
here are the pointers:
- the IT source: https://github.com/apache/maven-integration-testing/blob/master/core-it-suite/src/test/java/org/apache/maven/it/MavenITmng5771CoreExtensionsTest.java
- associated test projects: https://github.com/apache/maven-integration-testing/tree/master/core-it-suite/src/test/resources/mng-5771-core-extensions/client

I suppose writing the IT will simply require to copy testCoreExtension() method and not setting basedir on the Verifier, but add the new option

here is core-its intro: http://maven.apache.org/core-its/core-it-suite/

do you want to try? Or I do it (now that I had a look and found existing IT, seems easy)

**Comment by michael-o** [comment]
@hboutemy Thanks for helping out!

**Comment by hboutemy** [comment]
I know core ITs are intimidating: it took me years to really dig into the code!
This case seems a simple one (what I didn't really expect when thinking at it initially), since it's only improving existing MNG-5771 IT with just different launch conditions

if it is still too much intimidating, tell me: I'll do the IT showing the failure, and you'll add the fix :)

**Comment by rpatrick00** [comment]
I took a quick look at the MNG-5771 IT and it's not obvious to me exactly what I need to do so if you can add this in a few minutes, that is probably the best thing to do.  If not, I will try to look at it as I can and try some tests to figure out how it works...

**Comment by jvanzyl** [comment]
Please do not change an existing test. The conditions under which this IT passes should be left as-is and should continue running as it is in exactly the form it runs now. Making purely additional logic is fine, or copy that IT and make another test all together, do not change the logic or resources of the existing IT. It captures behavior that is currently deployed in released versions and so it should remain as a representation of that.

**Comment by hboutemy** [comment]
IT added https://github.com/apache/maven-integration-testing/commit/2e74409fc2593b4478c40f28c7fdabc60202c06b
no change on existing ITs, just added an additional run with different execution state: current dir = parent, mvn run with -f client/pom.xml
of course, it is currently failing: I let Tibor or Michael apply the fix

**Comment by hboutemy** [comment]
IT moved to its own suite in https://github.com/apache/maven-integration-testing/commit/3311f4e345fe7e6c343fae737631b727aba5b910 to avoid running this IT with Maven 3.3.x, as it will fail

**Comment by michael-o** [comment]
@hboutemy I am bit confused by the IT. The issue was about having `.mvn` properly located to expand `maven.config`. I would expect some system property to be added and retrieved. Am I wrong?

**Comment by hboutemy** [comment]
maybe I'm missing something...
{{.mvn}} directory has to be found by shell script because of {{jvm.config}} particularly
But once he directory is found by shell script, I suppose it is injected in JVM launch command to be used for every other file ({{maven.config}} and {{extensions.xml}}): that's perhaps where I'm wrong, I didn't check
I just looked at Linux shell update, that properly finds {{jvm.config}}: and I supposed that the IT checking for {{extensions.xml}} was consistent.
Perhaps this is just a beginning and has to be improved to check more that simply {{extensions.xml}}

**Comment by michael-o** [comment]
@rpatrick00 Can you shed some light here? What do you to properly picked up? `extensions.xml`, `maven.config`, and `jvm.config`, right?

**Comment by rpatrick00** [comment]
If you look at the shell script, it used to compute the "basedir" variable starting at the current working directory and then search upwards for the .mvn directory.  Then, it set the MAVEN_PROJECTBASEDIR environment variable and added the contents of jvm.config.  Finally, it passed it into the java command line using the -Dmaven.multiModuleProjectDirectory=${MAVEN_PROJECTBASEDIR} argument.

All my fix does it change the starting point to search for the .mvn directory if the -f/--file argument is specified.  The rest of the logic is exactly the same...

**Comment by hboutemy** [comment]
yes, that's exactly the part I was looking for. The shell script:

1. finds the basedir
2. uses it to load `jvm.config`
3. then injects the value as `maven.multiModuleProjectDirectory` property for the java part to use it to load `maven.config` and `extensions.xml`

then when the IT checks that `extensions.xml` is loaded, it shows one of the features: perhaps we could add checks for the 2 others, but I admint that when `extensions.xml` works, the other work too
(in fact, I didn't find any IT for the 2 others, then adding a test could be useful to fix lack of previous IT...)

**Comment by michael-o** [comment]
@hboutemy I would at least check for `jvm.config` because this is expanded in the script itself.

**Comment by hboutemy** [comment]
IMHO, the first thing is to check in the current fix, since ITs are currently broken
then improving the IT to check more parts could be interesting: priority 2

**Comment by hboutemy** [comment]
I just merged the patch
I had to change `==` tests to `=` since it did not work on Linux: the result is `if [ "$arg" = "-f" -o "$arg" = "--file" ]; then`
And it required a little update to the IT to force JVM fork, to force use of the shell script when launching IT build
Everything seems good: I hope ASF Jenkins will tell that it works like it worked on my machine :)

Thank you @rpatrick00 and Maven devs who helped him: this one is a great improvement

**Comment by rpatrick00** [comment]
@hboutemy, sorry but both "==" and "=" seem to work fine on Oracle Linux and MacOS (where I tested the sh script).  Not sure which version of Linux caused this but thanks for fixing it...

**Comment by hboutemy** [comment]
FYI I'm running Kubuntu 16.10
not a big issue: thank your for the great job
please take time to close the PR

**Comment by michael-o** [comment]
@hboutemy Are you waiting for the Jenkins build to close the JIRA issue? @rpatrick00 Thank you for the contribution.

**Comment by hboutemy** [comment]
yes: now that ASF Jenkins shows me it works not only on my Linux box but also on ASF Linux and on Windows, I just closed the Jira issue

**Comment by jira-importer** [comment]
Resolve #7889

**Comment by jira-importer** [comment]
Resolve #8062

**Pull request closed without merging**